### ISSUE TITLE
User Input guard/popup prevents further execution after stopping execution and kernel restarts

### ISSUE THREAD

**Issue by chiragjn** [jupyterlab/jupyterlab#10732]
## Description

It seems like an input guard with popup was added, while I can see it can be useful to catch user's attention, it is too disruptive for any other uses. For e.g. I use pdb `post_mortem` a lot to debug my expensive ML code to catch any bugs before I leave it for hours to run. But this new popup flow blocks and overlays itself over everything. Even if i dismiss it, it won't let me take any other action - not even stopping the cell or kernel restart can salvage the situation.

## Reproduce

```
jupyter==1.0.0
jupyter-client==6.1.7
jupyter-console==6.0.0
jupyter-contrib-core==0.3.3
jupyter-contrib-nbextensions==0.5.1
jupyter-core==4.7.0
jupyter-highlight-selected-word==0.2.0
jupyter-latex-envs==1.4.6
jupyter-nbextensions-configurator==0.4.1
jupyter-server==1.10.1
jupyterlab==3.1.0
jupyterlab-server==2.6.1
```

```python
import pdb

def my_very_expensive_function():
    a = "hello"
    b = 5
    c = a / b
    return c

try:
    my_very_expensive_function()
except Exception:
    pdb.post_mortem()
```

https://user-images.githubusercontent.com/10295418/127510749-d04b5413-3208-4e00-8643-7b4742eb6daa.mp4

## Expected behavior

Stopping cell execution should stop asking for input. Plus there should be an option to disable the popup behavior entirely

## Context

- Operating System and version:  Ubuntu 18.04
- Browser and version: Chrome 91
- JupyterLab version: 3.1.0

**Comment by geozeke** [comment]
Greetings!

First, thanks to the dev team for the amazing work on this project :-)

I chatted with a dev on Gitter, and he pointed me to [jupyterlab/jupyterlab#8713](https://github.com/jupyterlab/jupyterlab/pull/8713) when I asked about this change. It caught me by surprise too. The logic behind the change makes sense - to avoid blocking when attempting to re-run a code cell that's waiting for user input. It looks like it's set up to be a modal dialog, but you can click around it to get to the code cell / input box anyway.

It would be great to see an option to disable this in a custom user configuration in `tracker.jupyterlab-settings`. Maybe something like:

```
{"codeCellConfig": {
      "inputGuard": false
}}
```

**Comment by akhmerov** [comment]
A particularly disruptive situation is a `%debug` magic in IPython: there one is presented with the new input popup immediately after every interaction (since the debugger always waits for input).

**Comment by echarles** [comment]
What would be the solution to cover this case? This initial implementation (as shown on the image on https://github.com/jupyterlab/jupyterlab/pull/8713#issue-456135974) did not have a modal to ask for the input, but a modal to inform an input was already requested.

From the description and the tests I have run with your snippet, vemoving the model to ask the input won't help, right?

>  Even if i dismiss it, it won't let me take any other action - not even stopping the cell or kernel restart can salvage the situation.

Well I think this is the goal of the feature. You have to give an input before asking for a second input. In the ipdb case you mention, we fall in another completely case we have not take into account, true.

Unless we have a solution to cover this case on the UI, we can always fallback to the additional setting @geozeke has mentionned.

BTW I see the input prompt starts with ìpdb`. Another terrible hack would be to have a special case when the input prompt is that specific  ìpdb` string.

<img width="492" alt="Screenshot 2021-07-30 at 07 35 31" src="https://user-images.githubusercontent.com/226720/127605783-2143dc39-d996-4912-abea-de08b4b9da95.png">

WDYT?

**Comment by echarles** [comment]
> A particularly disruptive situation is a %debug magic in IPython: there one is presented with the new input popup immediately after every interaction (since the debugger always waits for input).

@akhmerov Can you give a bit more details on the issue you have faced with `%debug`? This will help to find a solution/workaround. Thx.

**Comment by chiragjn** [comment]
On re-tracing my steps I realised I was pressing Shift+Enter instead of Enter out of habit which was issue Run Cell instead of submitting input (0:08 - 0:09 in the video in the description). So pressing Enter works just fine. :) 
As far as I can tell the issue is in the JS state, specifically a cell stop and kernel restart does not remove the waiting flag from memory (a refresh of the page does reset that)


https://user-images.githubusercontent.com/10295418/127612163-89b786e0-6009-4fd9-92d9-7354e139dea0.mp4


In the video above,

First I issue a KeyboardInterrupt which works but till I don't submit the input element I can't do anything else, even though the KeyboardInterrupt has already stopped the process from waiting.
The second part is exactly the same with kernel restart.
The third part is a nitpick where I start an input, then clear all outputs, at this point nothing but a page refresh can fix things for me.

An option to disable the popup behaviour in the short term would be nice till contributors can figure out a good solution :)

**Comment by echarles** [comment]
> As far as I can tell the issue is in the JS state, specifically a cell stop and kernel restart does not remove the waiting flag from memory (a refresh of the page does reset that)

@chiragin Looks like this is another issue than the one raised at the beginning of this thread related to behaviour while debugging. If this is the case, can you open a separated issue for that?

**Comment by akhmerov** [comment]
> @akhmerov Can you give a bit more details on the issue you have faced with `%debug`? This will help to find a solution/workaround. Thx.

After rereading, I realized that I faced essentially the same issue as in the original post. To work with %debug one usually needs very short inputs interlaced with examining the frames. Now the best approach for the user is to remove the popup from the cell at each debugger step. While I wasn't confused by the new UI, I just found the pop-up very disruptive to the workflow.

Note that the popup by itself doesn't fix the problem of locking the kernel if the cell is reexecuted, especially if the user can press cancel or click away from the popup. That happens regardless of the popup now, and is a nice safeguard. Perhaps a notification would be more appropriate?

**Comment by echarles** [comment]
Maybe the next steps would be:

- Remove the modal that ask for user input to avoid breaking rapidly iterating users.
- Add a setting to disable the check for power users who know what they are doing.

**Comment by krassowski** [comment]
Off topic: @akhmerov @chiragjn have you tried JupyterLab debugger? Since you are already running JupyterLab 3.1, the only thing you need to do is tu update ipykernel:

```
pip install -U ipykernel
```

or

```
conda -c conda-forge update ipykernel
```

And a toggle should show up on the Notebook toolbar allowing you to enable it. Just curious what your thoughts are.

Of course it does not make the need for making the workflow less disruptive any less important. I fully agree that the modals should go away or at very least be made optional.

**Comment by akhmerov** [comment]
Hehe, I never saw the switch appear, even with jlab 3.1 and ipykernel 6.0.3 (both from conda-forge) :crying_cat_face:. Maybe now is the time to figure out what's wrong.

**Comment by echarles** [comment]
> Maybe now is the time to figure out what's wrong.

`pip install --upgrade --pre ipykernel` ?

**Comment by akhmerov** [comment]
```
anton@ottawa ~> pip install -U --pre ipykernel                                                                                                                                                      Requirement already satisfied: ipykernel in ./miniconda3/envs/py39/lib/python3.9/site-packages (6.0.3)
```

**Comment by meeseeksmachine** [comment]
This issue has been mentioned on **Jupyter Community Forum**. There might be relevant details there:

https://discourse.jupyter.org/t/missing-debugger-option/9851/12

**Comment by beckernick** [comment]
> A particularly disruptive situation is a `%debug` magic in IPython: there one is presented with the new input popup immediately after every interaction (since the debugger always waits for input).

I'm also running into this issue. It's particularly difficult as I often don't know what step I want to take until evaluating the new frame, which requires hitting cancel.

> Off topic: @akhmerov @chiragjn have you tried JupyterLab Debugger? Since you are already running JupyterLab 3.1, the only thing you need to do is to update ipykernel:

Thanks for the suggestion @krassowski ! I haven't explored this before, but will spend some time on it. From a quick test, this looks quite useful for when I am debugging a function that I've defined in scope. For example, I can step through `func`:

![Screen Shot 2021-07-30 at 10 36 37 AM](https://user-images.githubusercontent.com/8457388/127669239-433d6047-364c-495d-80c7-48298d6c71a4.png)

But it's not as clear to me how I can step through a function defined elsewhere. Concretely, I'm trying to step through an unexpected failure in library code but the first step brings me to the final frame and the error. This is my first time using this, so I may be missing something obvious.

![Screen Shot 2021-07-30 at 10 39 34 AM](https://user-images.githubusercontent.com/8457388/127669586-f61ca0c8-eb3c-431c-ac09-c592104d0cc2.png)

Using %debug or %pdb, i can explicitly step through the frames.

**Comment by krassowski** [comment]
> But it's not as clear to me how I can step through a function defined elsewhere. Concretely, I'm trying to step through an unexpected failure in library code but the first step brings me to the final frame and the error. This is my first time using this, so I may be missing something obvious.

There is a draft PR for that: https://github.com/jupyterlab/jupyterlab/pull/10298 and it might make it into 3.2 - please do subscribe to that one and help testing once ready :)

**Comment by beckernick** [comment]
Nice! Thanks for the pointer. Happy to take a look and test when ready.

**Comment by chiragjn** [comment]
> have you tried JupyterLab Debugger? Since you are already running JupyterLab 3.1, the only thing you need to do is to update ipykernel:

I did read about it and I would love to, unfortunately, I have to use Python 3.6 at the moment and ipykernel 6.x is only available for 3.7+. So maybe someday in the future :sweat_smile:

**Comment by mlucool** [comment]
The restart state looks more like a bug than something to fix with user setting. As you can see here, it's working well until you restart, then the notebook becomes unusable.

![lock](https://user-images.githubusercontent.com/1813603/127933847-4fadf55a-3560-45df-aa47-754d7e630e2d.gif)

**Comment by echarles** [comment]
Rereading this issue, there may be different issues reported:

- The modal to get input which is annoying
- The usage of ipdb which should not block multiple input requests? - I am not clear on this one.
- The usage of %debug which should not block multiple input requests? - I am not clear on this one.
- The issue mentioned in the title where the blocking flag should be reset ofter restart.

**Comment by wsfreund** [comment]
> The modal to get input which is annoying

My point concerns only this one: getting this prompt every time the user interacts with ipdb reduces the speed he interacts with the code and, hence, the productivity.

Users should be able to disable such dialog if they wish to do so, and I even dare to say that perhaps it would be better if it is disabled by default for debugging prompts.... but I am not sure how difficult it would be to implement such feature.

Particularly, I would not enable such dialog anyway, because it does not have much use case for my day to day usage.

**Comment by jasongrout** [comment]
Another discussion about this from @fperez is at https://gitter.im/jupyterlab/jupyterlab?at=610a1f36933ee46b753babb2

**Comment by fperez** [comment]
Thanks @jasongrout for pointing me here, indeed I think this should be reverted so we can explore a cleaner solution... Even with the new debugger, many of us use `%debug` reflexively, and having a pop-up in the middle of that is super jarring.

I also listed on the gitter thread other issues with things like audio control, in addition to how easy it is to render the notebook unusable after interrupting the kernel...

I appreciate all the thought that went into the original idea and that handling blocking input in an otherwise highly asynchronous experience is a minefield, but I feel pretty strongly this should be reverted before we cause more disruption to users (esp. given all the excitement to try out 3.1 with its many awesome features). That would give us time to explore a better solution slowly...

**Comment by jasongrout** [comment]
Here's another idea for dealing with the original problem at https://github.com/jupyterlab/jupyterlab/issues/8632: revert   #8713 and introduce a new kernel (pseudo-)state that is "waiting for input", shown as a kernel status like busy, idle, etc. If the user loses the input box that comes up in the output region of a cell like explained in #8632, they can click on the "waiting for input" status in the notebook toolbar or the status bar to pop up the modal that was introduced in #8713.

**Comment by fperez** [comment]
I like @jasongrout's idea... While it may require a bit more user interaction and is perhaps slightly less discoverable, I do find it preferable to a fully blocking modal dialog.  If people still get a little confused, we could consider stronger visual cues, such as styling the in-notebook input box with a more visible border/highlight, and/or coloring the status bar kernel indicator in these cases with a highlight... 

These are just ideas, all to suggest that for this issue, I think we'll get more mileage out of providing indicators to the user of what's going on with the kernel (blocking on input instead of busy on execution).

**Comment by echarles** [comment]
https://github.com/jupyterlab/jupyterlab/pull/8713 had two things:

1. A modal that says "Waiting on input - Please submit your input before running again the cell" which is the gist of the fix for the issue - This was the initial implementation but in the discussion on https://github.com/jupyterlab/jupyterlab/pull/8713 , we came with the second modal.
2. A modal with the exact same input text field as the one which is normally displayed. This second modal is indeed optional and should be removed.

If we remove the first modal, we gonna suffer back from https://github.com/jupyterlab/jupyterlab/issues/8632 (When the kernel is waiting for input, it can get deadlocked). I understand that the kernel status icon "waiting for user input" and a more visible inputbox may gain user attention, but at the end, if the user generates a second input while the first one is not yet honored, the kernel will be deadlocked.

So far, a lot of users have complained about the second modal, but it is not clear to me if the first modal is also an issue and should be removed.

**Comment by fperez** [comment]
Thx @echarles for the clear description! I agree that the first modal isn't nearly as problematic - my only issue with it is that, as noted above, I've seen it _mistriggered_. If we can resolve that, then I see no problem with keeping it...

The misfire seems to arise if the kernel is interrupted while waiting on a loop say like

```python
for i in range(3):
  input(f'Hi, enter something:')
```

If you interrupt the kernel and then try to run a new cell, you get the blocking dialog:

![image](https://user-images.githubusercontent.com/57394/128198778-e67663f7-f4ae-4be4-9a77-874d603341ea.png)

In that case, the `input` loop had already been stopped, so the dialog should not pop up.

I think if that case is resolved, we can restore modal 1 and get rid of modal 2.

**Comment by echarles** [comment]
> In that case, the input loop had already been stopped, so the dialog should not pop up.
> I think if that case is resolved, we can restore modal 1 and get rid of modal 2.

Thx a lot @fperez for bringing on this case. I have read another case where the popup 1 was shown after kernel restart (if you restart, any new input should be allowed).

I will review those cases and make sure modal 1 is not *mistriggered* and will get rid of modal 2 in all cases.

**Comment by jasongrout** [comment]
Another problem with global modals: What happens if you have two notebook tabs A and B simultaneously executing, and notebook B has an input requested? How do you know it is for notebook B, and not A? What if B is in a tab behind A, so you can't even see B or its context?

One thought is that we have tab-specific modals (similar to how we have application-specific modals on macOS). But I think better is to just have a way to get a kernel unstuck from requesting input like suggested [above](https://github.com/jupyterlab/jupyterlab/issues/10732#issuecomment-892376892), with perhaps better styling like [Fernando suggested](https://github.com/jupyterlab/jupyterlab/issues/10732#issuecomment-892383175)

**Comment by echarles** [comment]
@jasongrout Are you saying to completely revert https://github.com/jupyterlab/jupyterlab/pull/8713, removing also the guard? In that case, the user risks to block his kernel if he requests a second input.

**Comment by jasongrout** [comment]
> @jasongrout Are you saying to completely revert #8713, removing also the guard? In that case, the user risks to block his kernel if he requests a second input.

Which just puts us back to the status quo up through 3.0. The question (for me) is about which behavior is more disruptive - the chance that a user deadlocks their kernel (which we've always had up to this point), or the disruption we're seeing with the modal.

From https://github.com/jupyterlab/jupyterlab/issues/10732#issuecomment-889842485 and https://github.com/jupyterlab/jupyterlab/issues/10732#issuecomment-889845093 above, the conclusion seemed to be that we should revert even the modal asking for input as it was too disruptive for a common usecase.

**Comment by echarles** [comment]
> From #10732 (comment) and #10732 (comment) above, the conclusion seemed to be that we should revert even the modal asking for input as it was too disruptive for a common usecase.

Maybe you mean "... revert even the modal saying Waiting on input - Please submit your input before running again the cell" ?

**Comment by jasongrout** [comment]
> Maybe you mean "... revert even the modal saying Waiting on input - Please submit your input before running again the cell"? (which is Modal 1)

Thanks for being generous in your interpretation. I was getting confused between the two modals. It sounds like your comment far above:

> Remove the modal that ask for user input to avoid breaking rapidly iterating users.

and your other comment more recently above:

> I will review those cases and make sure modal 1 is not mistriggered and will get rid of modal 2 in all cases.

are calling for the same thing. +1 to your plan in https://github.com/jupyterlab/jupyterlab/issues/10732#issuecomment-892709812 for the immediate future.

I still think using modals is probably not a good long-term solution, but the modal you are proposing to leave in does not sound like it is causing big disruptions from user reports (as long as it is not mistriggered, like you say).

Thanks again for following up on this issue.

**Comment by echarles** [comment]
> are calling for the same thing. +1 to your plan in #10732 (comment) for the immediate future.
> I still think using modals is probably not a good long-term solution, but the modal you are proposing to leave in does not sound like it is causing big disruptions from user reports (as long as it is not mistriggered, like you say).
> Thanks again for following up on this issue.

Sounds good and sounds like a plan. Thx you for your attention on this.

I will start working towards a PR tomorrow and hope to find ways to fix the mistriggered modal 1 (the good thing is that we have reproducers for that).

**Comment by fperez** [comment]
Thanks for all the work on this, folks... My gut tells me that perhaps there's a good solution to be found with no modals at all, but given that a) @echarles seems reasonably convinced the misfires with the simpler dialog can be managed, b) it does address another nasty failure mode; c) it hasn't generated nearly as many concerns; I'm for now happy to try a "minimal reversion" as Eric suggests. 

Once that lands we should keep exploring whether even that dialog does still cause any unforeseen problems, but we can take it now one step at a time...

**Comment by jasongrout** [comment]
Given the user reports and disruption, can we target this to 3.1.x, instead of waiting 2 months for 3.2?

**Comment by fperez** [comment]
Yes, please! I'm holding back on updating our working research hub to 3.1 b/c of this. For me, 3.1 is basically not usable, I'm afraid, due to this issue.

**Comment by geozeke** [comment]
+1

Please from me, too! Classes start on August 23rd. Would love to see this rolled back before then :-)

**Comment by fperez** [comment]
Good point, @geozeke!! I'd forgotten about that part - I was talking about a small research hub above, but I need to settle on our image for our 1,000+ person [Data 100](https://ds100.org) course that also starts very soon! And I can't use 3.1 in its current form, unfortunately.

**Comment by echarles** [comment]
Sorry to read that! It sounds like there is enough emergency here to completely revert and ensure we don't introduce edge behavior for the users.

I have opened https://github.com/jupyterlab/jupyterlab/pull/10779 to revert on 3.1.x branch which I think is the one upon which the next 3.1 release will be done with this fix (cc/ @blink1073).

I guess the revert should be ported to master. Then of course, I can start with a new PR for 3.2 and master which only implements Modal 1, taking into account the reported cases. From there, I hope enough users/developers will test that branch before any merge to ensure no regression is introduced.

**Comment by fperez** [comment]
No worries @echarles, your rapid response is much appreciated. And you were trying to solve a tricky problem! Unforeseen side effects of a solution do happen, and we just iterate and keep improving, no worries :) 

With a rapid update for a 3.1.x release with this fixed, 3.1 will be awesome, and we can continue testing better approaches to this. We now know of a number of edge cases to watch for, so that's progress. It will make the next attempt easier to work through.

**Comment by echarles** [comment]
fyi the complete revert is merged in 3.1.x and master branch. It is now a matter to cut the next 3.1.x release so you can use it.

**Comment by fperez** [comment]
On Sat, Aug 7, 2021 at 10:12 PM Eric Charles ***@***.***>
wrote:

> fyi the complete revert is merged in 3.1.x and master branch. It is now a
> matter to cut the next 3.1.x release so you can use it.
>
Thank you so much Eric for your diligence on this one, much appreciated!

**Comment by blink1073** [comment]
Fix is released: https://github.com/jupyterlab/jupyterlab/releases/tag/v3.1.5

**Comment by fperez** [comment]
Huge thanks all! Should we close this one?

**Comment by echarles** [comment]
I am working on a PR for 3.2 for Modal 1 and plan to link this PR to this issue. There is very good reproducer logged here, so I would prefer keep it open until we have a decision for 3.2.

**Comment by fperez** [comment]
np @echarles  - feel free to ping if you'd like more testing/input once the PR goes up, happy to take a look!

**Comment by echarles** [comment]
> np @echarles - feel free to ping if you'd like more testing/input once the PR goes up, happy to take a look!

Thx. I have opened https://github.com/jupyterlab/jupyterlab/pull/10792 and would like more user tests to identify any issue with the implementation. Compared to the previous attempt (1) the second modal asking for user input is not more shown (2) the interrupt, shutdown and restart kernel resets the guard.

Looking forward feedback from any contributor to this thread or anyone interested.

**Comment by echarles** [comment]
> np @echarles - feel free to ping if you'd like more testing/input once the PR goes up, happy to take a look!

@fperez I understand you are not asking party for this feature and that you have so many other things to do, but as you proposed your help, feel free to test your typical workflow case and report any feedback your would have here.

The PR has been tested by @goanpeca who has told me it was working fine.

The link to the binder is https://hub-binder.mybinder.ovh/user/datalayer-externals-jupyterlab-qhs03xw7/lab-dev/lab

Thx again!

**Comment by mlucool** [comment]
> Appreciate if anyone can try this branch on the following binder link https://mybinder.org/v2/gh/datalayer-externals/jupyterlab/feat%2Finput-guard-2?urlpath=lab-dev and see if the implementation regarding the double input management fits their need.

I did a quick test and it looks good to me.

**Comment by echarles** [comment]
Thx a lot @fperez and @mlucool for your time and feedbacks. Happy to see the new PR doesn't break the normal flow of data analysts when using jupyterlab and debugging their code.

**Comment by echarles** [comment]
https://github.com/jupyterlab/jupyterlab/pull/1079 is merged in 3.1.x branch. I am closing now this issue.

Thank you all for the reports and constructive discussions.

Please reopen this (or a new issue) if you encounter more troubles with the recent code merge.
